src/extensions/step_handlers/script_handler/src/script_handler.cpp

```diff
@@ -134,25 +139,76 @@ static ADUC_Result Script_Handler_DownloadPrimaryScriptFile(ADUC_WorkflowHandle
     workflowId = workflow_peek_id(handle);
     workFolder = workflow_get_workfolder(handle);
 
-    createResult = ADUC_SystemUtils_MkSandboxDirRecursive(workFolder);
-    if (createResult != 0)
+    // Construct the primary script file path
+    filePath << workFolder << "/" << scriptFileName;
+    scriptFilePath = filePath.str();
+
+    if (!ADUC_HashUtils_GetShaVersionForTypeString(
+            ADUC_HashUtils_GetHashType(entity->Hash, entity->HashCount, 0), &algVersion))
     {
-        Log_Error("Unable to create folder %s, error %d", workFolder, createResult);
-        result = { ADUC_Result_Failure, ADUC_ERC_SCRIPT_HANDLER_CREATE_SANDBOX_FAILURE };
+        Log_Error(
+            "FileEntity for %s has unsupported hash type %s",
+            scriptFilePath.c_str(),
+            ADUC_HashUtils_GetHashType(entity->Hash, entity->HashCount, 0));
+        result.ExtendedResultCode = ADUC_ERC_CONTENT_DOWNLOADER_FILE_HASH_TYPE_NOT_SUPPORTED;
         goto done;
     }
 
-    try
+    // If file exists and has a valid hash, then skip download.
+    // Otherwise, delete an existing file, then download.
+    Log_Debug("Check whether '%s' has already been download into the work folder.", scriptFilePath.c_str());
+
+    if (access(scriptFilePath.c_str(), F_OK) == 0)
     {
-        ExtensionManager_Download_Options downloadOptions = {
-            .retryTimeout = DO_RETRY_TIMEOUT_DEFAULT,
-        };
+        char* hashValue = ADUC_HashUtils_GetHashValue(entity->Hash, entity->HashCount, 0 /* index */);
+        if (hashValue == nullptr)
+        {
+            result = { .ResultCode = ADUC_Result_Failure,
+                       .ExtendedResultCode = ADUC_ERC_CONTENT_DOWNLOADER_INVALID_FILE_ENTITY_NO_HASHES };
+            goto done;
+        }
+
+        // If target file exists, validate file hash.
+        // If file is valid, then skip the download.
+        bool validHash =
+            ADUC_HashUtils_IsValidFileHash(scriptFilePath.c_str(), hashValue, algVersion, false /* suppressErrorLog */);
+
+        if (!validHash)
+        {
+            // Delete existing file.
+            if (remove(scriptFilePath.c_str()) != 0)
+            {
+                Log_Error("Cannot delete existing file that has invalid hash.");
+                result.ExtendedResultCode = ADUC_ERC_CONTENT_DOWNLOADER_CANNOT_DELETE_EXISTING_FILE;
+                goto done;
+            }
+        }
 
-        result = ExtensionManager::Download(entity, handle, &downloadOptions, nullptr);
+        fileExist = true;
     }
-    catch (...)
+
+    if (!fileExist)
     {
-        result.ExtendedResultCode = ADUC_ERC_SCRIPT_HANDLER_DOWNLOAD_PRIMARY_FILE_FAILURE_UNKNOWNEXCEPTION;
+        createResult = ADUC_SystemUtils_MkSandboxDirRecursive(workFolder);
```

Comment:
I feel that work folder should be created by extension manager download; otherwise, every content handler has to dup logic to create it before calling download.